### PR TITLE
WIP: Support delta update on macOS

### PR DESCRIPTION
**Pull request by simonhong**
EdDSA public key should be set in .npmrc like sparkle_eddsa_key=XXXXX.
This key can be generated by using generated `out/XXX/generate_keys`. It'll create private key and store in keychain.

Resolves https://github.com/brave/brave-browser/issues/6555

Old app path should be passed via `last_chrome_installer` like below. It's relative path from output dir(ex, src/out/Release).
Then, `BinaryDelta` will create `patch.delta`.
npm run create_dist Release --  --build_delta_installer --last_chrome_installer=delta/Brave\ Browser.app

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: build/mac/build_delta_installer.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import sys
+
+
+def Main(argv):
+  parser = optparse.OptionParser('%prog [options]')
+  parser.add_option('--binary-delta', dest='binary_delta_path', action='store',
+      type='string', default=None, help='The path of BinaryDelta binary.')
+  parser.add_option('--sign-update', dest='sign_update_path', action='store',
+      type='string', default=None, help='The path of sign_update binary')
+  parser.add_option('--old-app', dest='old_app_path', action='store',
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:18]: let's use `current` and `previous` vs `new` and `old`

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

File: build/mac/build_delta_installer.py (hunk 1 above)
Comment (mihaiplesa) [reply build/mac/build_delta_installer.py:18]: in fact it might not be the current/latest version so it makes sense how it is now

File: DEPS [hunk 2]
```diff
@@ -8,7 +8,7 @@ deps = {
   "vendor/boto": "https://github.com/boto/boto@f7574aa6cc2c819430c1f05e9a1a1a666ef8169b",
   "vendor/python-patch": "https://github.com/brave/python-patch@d8880110be6554686bc08261766538c2926d4e82",
   "vendor/omaha":  "https://github.com/brave/omaha.git@cd2499ce23b2f8c532e1c504adf41d3e946780d2",
-  "vendor/sparkle": "https://github.com/brave/Sparkle.git@a8ba961ee53e7bf8484845ec14050b74f530e813",
+  "vendor/sparkle": "https://github.com/brave/Sparkle.git@72a4d6974083301c3e182940dbf36f8c744156bb",
```
Comment (mherrmann) [inline DEPS:11]: Commit 72a4d6 "does not belong to any branch" in the brave/Sparkle repository ([source](https://github.com/brave/Sparkle/commit/72a4d6974083301c3e182940dbf36f8c744156bb)). Is this intended?

Comment (simonhong) [reply DEPS:11]: Well, not sure why it says like that. However, this is WIP PR and merge commit hash will be used when sparkle PR is merged.

File: build/mac/build_delta_installer.py (hunk 1 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py:18]: So, it's ok to use `old` and `new`?

Comment (mihaiplesa) [reply build/mac/build_delta_installer.py:18]: yes, we can keep

File: build/mac/reverse_framework_version_symlink.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import subprocess
+import shutil
+import sys
+
+
+def Main(argv):
+  parser = optparse.OptionParser('%prog [options]')
+  parser.add_option('--version-directory-path', dest='version_directory', action='store',
+      type='string', default=None, help='The path of BinaryDelta binary.')
+  parser.add_option('--version', dest='version', action='store',
```
Comment (mihaiplesa) [inline build/mac/reverse_framework_version_symlink.py:20]: is the `help` text accurate?

File: build/mac/build_delta_installer.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  # generate patch file
+  command = [options.binary_delta_path, 'create', options.old_app_path, options.new_app_path, options.output_path]
+  try:
+      subprocess.check_call(command)
+  except subprocess.CalledProcessError as e:
+      print(e.output)
+      raise e
+
+  # sign patch file
+  command = [options.sign_update_path, options.output_path]
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:39]: we should add the option to pass `-s` which takes the private key value

Comment (mihaiplesa) [reply build/mac/build_delta_installer.py:39]: https://github.com/sparkle-project/Sparkle/blob/master/sign_update/main.swift#L67

File: build/mac/build_delta_installer.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import optparse
+import subprocess
+import sys
+
+
+def Main(argv):
+  parser = optparse.OptionParser('%prog [options]')
+  parser.add_option('--binary-delta', dest='binary_delta_path', action='store',
+      type='string', default=None, help='The path of BinaryDelta binary.')
+  parser.add_option('--sign-update', dest='sign_update_path', action='store',
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:16]: should add option to pass signing key here instead of expecting it to be in the keychain

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

File: build/mac/build_delta_installer.py (hunk 4 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py:39]: Hmm, `-s` is not supported with our sparkle version.
Need to cherry-pick https://github.com/sparkle-project/Sparkle/commit/531b767c0b070acc5ef544c7f77eee28e0344dcb#diff-ff51bfaabc34dea371c9dc48b1d67cde73bda2d66ae327bce44ba325ad28d1fb

Comment (simonhong) [reply build/mac/build_delta_installer.py]: Fixed.

File: build/mac/reverse_framework_version_symlink.py (hunk 3 above)
Comment (simonhong) [reply build/mac/reverse_framework_version_symlink.py:20]: Oop, fixed.

File: build/mac/build_delta_installer.py (hunk 5 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py:16]: Done.
`.npmrc` should have below two key.
```
sparkle_eddsa_public_key=xxxxxxxxxxxxxxxxxxx
sparkle_eddsa_private_key=xxxxxxxxxxxxxxxxxxxxx
```

File: build/commands/lib/config.js [hunk 6]
```diff
@@ -1618,9 +222,8 @@ Config.prototype.buildArgs = function () {
     enable_hangout_services_extension: this.enable_hangout_services_extension,
     enable_cdm_host_verification: this.enableCDMHostVerification(),
     skip_signing: !this.shouldSign(),
-    sparkle_eddsa_public_key: this.sparkleEdDSAPublicKey,
-    sparkle_eddsa_private_key: this.sparkleEdDSAPrivateKey,
```
Comment (simonhong) [inline build/commands/lib/config.js]: @mihaiplesa Why did you delete private key? We need it for eddsa signing for patch and dmg file.

Comment (simonhong) [reply build/commands/lib/config.js]: I reverted this change by https://github.com/brave/brave-core/pull/6846/commits/c61bb41e9430432f908436fbdec0f4978055b3f7
I assumed you deleted this accidently.

File: build/mac/build_delta_installer.py [hunk 7]
```diff
@@ -21,29 +21,47 @@ def Main(argv):
...
+  command = [options.sign_update_path, '-s', options.sign_key, options.patch_output_path]
   try:
-      subprocess.check_call(command)
+      subprocess.check_call(command, stdout=file)
+  except subprocess.CalledProcessError as e:
+      print(e.output)
+      raise e
+  file.close()
+
+  # sign dmg
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:56]: we should sign the dmg even when not building the delta (if the sparkle private key is set)

Comment (mihaiplesa) [reply build/mac/build_delta_installer.py:56]: same for dsa, if the private key is passed in, we should sign the dmg

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

File: build/commands/lib/config.js [hunk 8]
```diff
@@ -106,6 +106,8 @@ const Config = function () {
   this.mac_installer_signing_identifier = getNPMConfig(['mac_installer_signing_identifier']) || ''
   this.mac_signing_keychain = getNPMConfig(['mac_signing_keychain']) || 'login'
   this.mac_signing_output_prefix = 'signing'
+  this.sparkleEdDSAPublicKey = getNPMConfig(['sparkle_eddsa_public_key']) || ''
+  this.sparkleEdDSAPrivateKey = getNPMConfig(['sparkle_eddsa_private_key']) || ''
```
Comment (mihaiplesa) [inline build/commands/lib/config.js:110]: can add here sparkleDSAPrivateKey for signing with DSA if present

File: BUILD.gn [hunk 9]
```diff
@@ -241,9 +241,19 @@ group("create_dist") {
...
     } else {
-      deps += [
-        "build/mac:finalize_dmg_pkg"
-      ]
+      deps += [ "build/mac:finalize_dmg_pkg" ]
+
+      if (sparkle_eddsa_private_key != "") {
+        deps += [ "build/mac:generate_eddsa_sig_for_dmg" ]
+
+        if (build_delta_installer) {
```
Comment (mihaiplesa) [inline BUILD.gn:249]: building delta shouldn't depend on private key but signing delta should

File: BUILD.gn [hunk 10]
```diff
@@ -241,9 +241,19 @@ group("create_dist") {
...
+          deps += [ "build/mac:build_delta_installer" ]
```
Comment (mihaiplesa) [inline BUILD.gn:250]: to extract delta signing into `mac:generate_eddsa_sig_for_delta` which calls `generate_eddsa_sig.py`

File: BUILD.gn [hunk 11]
```diff
@@ -241,9 +241,19 @@ group("create_dist") {
...
+        }
+      }
+
+      if (sparkle_dsa_private_key_file != "") {
```
Comment (mihaiplesa) [inline BUILD.gn:254]: minor, this goes up (alphabetically)

File: build/commands/lib/config.js [hunk 12]
```diff
@@ -106,6 +106,9 @@ const Config = function () {
...
+  this.sparkleDSAPrivateKeyFile = getNPMConfig(['sparkle_dsa_private_key_file']) || ''
```
Comment (mihaiplesa) [inline build/commands/lib/config.js:111]: 2 up

File: build/commands/lib/config.js [hunk 13]
```diff
@@ -222,6 +225,9 @@ Config.prototype.buildArgs = function () {
     enable_cdm_host_verification: this.enableCDMHostVerification(),
     skip_signing: !this.shouldSign(),
     chrome_pgo_phase: this.chromePgoPhase,
+    sparkle_eddsa_public_key: this.sparkleEdDSAPublicKey,
+    sparkle_eddsa_private_key: this.sparkleEdDSAPrivateKey,
+    sparkle_dsa_private_key_file: this.sparkleDSAPrivateKeyFile,
```
Comment (mihaiplesa) [inline build/commands/lib/config.js:230]: 2 up

File: build/config.gni [hunk 14]
```diff
@@ -33,13 +35,13 @@ declare_args() {
...
 
-  # Used for generating delta installer on Win.
+  # Used for generating delta installer on Win/macOS.
   build_delta_installer = false
   last_chrome_installer = ""
 }
 
 if (base_sparkle_update_url == "") {
-  base_sparkle_update_url = "https://updates.bravesoftware.com/sparkle/Brave-Browser"
+  base_sparkle_update_url = "https://updates.brave.com/sparkle/Brave-Browser"
```
Comment (mihaiplesa) [inline build/config.gni:44]: we have to remember to revert this

File: build/mac/BUILD.gn [hunk 15]
```diff
@@ -128,6 +128,104 @@ if (skip_signing) {
...
+      "$target_dmg_eddsa_path",
+    ]
+
+    deps = [
+      "//brave/vendor/sparkle:build_sparkle_framework",
+      ":copy_dmg_pkg",
+    ]
+  }
+
+  action("generate_dsa_sig_for_dmg") {
```
Comment (mihaiplesa) [inline build/mac/BUILD.gn:161]: this goes up

File: build/mac/build_delta_installer.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  except subprocess.CalledProcessError as e:
+      print(e.output)
+      raise e
+
+
+def Main(argv):
+  parser = optparse.OptionParser('%prog [options]')
+  parser.add_option('--binary-delta', dest='binary_delta_path', action='store',
+      type='string', default=None, help='The path of BinaryDelta binary.')
+  parser.add_option('--sign-update', dest='sign_update_path', action='store',
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:39]: this goes out once signing is extracted to GN

File: build/mac/build_delta_installer.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+      type='string', default=None, help='The path of sign_update binary')
+  parser.add_option('--sign-key', dest='sign_key', action='store',
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:41]: ditto

File: build/mac/build_delta_installer.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+      type='string', default=None, help='The private key to sign patch file')
+  parser.add_option('--root-out-dir', dest='root_out_dir_path', action='store',
+      type='string', default=None, help='The path of root output dir.')
+  parser.add_option('--old-dmg', dest='old_dmg_path', action='store',
+      type='string', default=None, help='The path of old dmg.')
+  parser.add_option('--new-dmg', dest='new_dmg_path', action='store',
+      type='string', default=None, help='The path of new dmg.')
+  parser.add_option('--patch-output', dest='patch_output_path', action='store',
+      type='string', default=None, help='The path of generated delta file.')
+  parser.add_option('--patch-eddsa-output', dest='patch_eddsa_output_path', action='store',
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:51]: ditto

File: build/mac/build_delta_installer.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+  # generate patch file
+  print('-> generate patch file with ' + old_app_path + ' and ' + new_app_path)
+  command = [options.binary_delta_path, 'create', old_app_path, new_app_path, options.patch_output_path]
+  try:
+      subprocess.check_call(command)
+  except subprocess.CalledProcessError as e:
+      print(e.output)
+      raise e
+
+  # sign patch file
```
Comment (mihaiplesa) [inline build/mac/build_delta_installer.py:76]: this moves to GN

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

File: BUILD.gn (hunk 9 above)
Comment (simonhong) [reply BUILD.gn:249]: got it.

File: BUILD.gn (hunk 10 above)
Comment (simonhong) [reply BUILD.gn]: done.

File: BUILD.gn (hunk 11 above)
Comment (simonhong) [reply BUILD.gn:253]: done.

File: build/commands/lib/config.js (hunk 12 above)
Comment (simonhong) [reply build/commands/lib/config.js]: done.

File: build/commands/lib/config.js (hunk 13 above)
Comment (simonhong) [reply build/commands/lib/config.js]: done.

File: build/mac/BUILD.gn (hunk 15 above)
Comment (simonhong) [reply build/mac/BUILD.gn]: done.

File: build/mac/build_delta_installer.py (hunk 16 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py]: removed.

File: build/mac/build_delta_installer.py (hunk 17 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py]: removed.

File: build/mac/build_delta_installer.py (hunk 18 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py]: removed.

File: build/mac/build_delta_installer.py (hunk 19 above)
Comment (simonhong) [reply build/mac/build_delta_installer.py]: removed.

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

File: build/mac/BUILD.gn [hunk 20]
```diff
@@ -128,6 +128,127 @@ if (skip_signing) {
...
+      "$target_delta_eddsa_path",
+    ]
+
+    deps = [
+      "//brave/vendor/sparkle:build_sparkle_framework",
+      ":build_delta_installer",
+    ]
+  }
+
+  action("build_delta_installer") {
```
[the remark this answers is not in the document]
Comment (simonhong) [reply build/mac/BUILD.gn]: done.

**Review by mihaiplesa — APPROVED** [review]

**Comment by bsclifton** [comment]
Updated to add milestone. Great job all! 😄
